icu4c/source/common/loclikely.cpp

```diff
@@ -1243,8 +1241,16 @@ _ulocimp_addLikelySubtags(const char* localeID,
                           UErrorCode* status) {
     char localeBuffer[ULOC_FULLNAME_CAPACITY];
 
-    if (do_canonicalize(localeID, localeBuffer, sizeof localeBuffer, status)) {
-        return _uloc_addLikelySubtags(localeBuffer, sink, status);
+    int32_t canonicalizedSize = do_canonicalize(localeID, localeBuffer, sizeof localeBuffer, status);
+    if (*status == U_BUFFER_OVERFLOW_ERROR || *status == U_STRING_NOT_TERMINATED_WARNING) {
+        icu::LocalMemory<char> heapLocaleBuffer((char*)uprv_malloc(canonicalizedSize + 2));
```

Comment:
Hmm... if `uprv_malloc` fails to allocate memory (due to OOM), then I'm not sure if `status` will be set to `U_MEMORY_ALLOCATION_ERROR`.
It doesn't look like either `do_canonicalize` or `uloc_canonicalize` check that the input buffer is non-NULL, so it might need to be checked here beforehand (with `heapLocaleBuffer.isNull()`) and the `status` set.
(Also, same with the lines 1302-1304 below as well).

Reply:
nit: maybe consider a C++ style cast since this file a .cpp? (Though to be fair, the rest of the file doesn't use them, so perhaps moot).

Reply:
agree with jeff about checking the null return of malloc, we surely should check that and set to U_MEMORY_ALLOCATION_ERROR if the return of malloc is nullptr

Reply:
Good catch(es).  Done.

It makes for pretty ugly code.  I thought about going to the extra trouble of factoring all this logic out into some kind of utility object, but it didn't seem worth the trouble for just these two (well, three) call sites.  I could do that if you guys insisted, but I can't help thinking it'd be better to wait until we see we need this kind of thing in more places.

Reply:
Thanks!

Hmm.. I don't know if it is worth factoring it out into a separate helper for now. This seems fine for the 2/3 call sites.

But thinking about it some more:
It makes it wonder if maybe we'd want something like `LocalPointer`'s [`adoptInsteadAndCheckErrorCode`](https://github.com/unicode-org/icu/blob/main/icu4c/source/common/unicode/localpointer.h#L319), but for the `LocalMemory` class.

`LocalMemory` already has [`adoptInstead`](https://github.com/unicode-org/icu/blob/main/icu4c/source/common/cmemory.h#L243)  -- so I think adding a `adoptInsteadAndCheckErrorCode` wouldn't be out of place for the class... (But that probably wants to be with a different ticket for discussion...)

Reply:
After thinking it over, I decided the best way to address both your and Frank's concerns _was_ to add a helper class.  I hope you agree.  I personally think It makes the code look a lot simpler.  I'm not sure ulocimp.h was the right place to put the helper class; if you all have a better idea, let me know.